test/Test/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,525 @@
+appendedKeys (AppendSinglePage _ lastKey)
+    = [lastKey]
+appendedKeys (AppendMultiPage key overflowPageCount)
+    = key : genericReplicate overflowPageCount key
+
+{-
+    Yields the index that results from performing a sequence of append
+    operations, starting with no keys.
+-}
+indexFromAppends :: [Append] -> IndexOrdinary
+indexFromAppends appends = indexFromLastKeys (concatMap appendedKeys appends)
+
+{-
+    Yields the serialized key list that results from performing a sequence of
+    append operations, starting with no keys.
+-}
+lastKeysBlockFromAppends :: [Append] -> Primitive.Vector Word8
+lastKeysBlockFromAppends appends = lastKeysBlock where
+
+    lastKeys :: [SerialisedKey]
+    lastKeys = concatMap appendedKeys appends
+
+    lastKeysBlock :: Primitive.Vector Word8
+    lastKeysBlock = Primitive.concat (lastKeysBlocks lastKeys)
+
+{-
+    Incrementally constructs an index, using the functions 'new', 'append', and
+    'unsafeEnd'.
+-}
+incrementalConstruction :: [Append] -> (IndexOrdinary, Primitive.Vector Word8)
+incrementalConstruction appends = runST $ do
+    acc <- new keyCount minChunkSize
+    commonChunks <- mapM (flip append acc) appends
+    (remnant, unserialised) <- unsafeEnd acc
+    let
+
+        serialised :: Primitive.Vector Word8
+        serialised = Primitive.concat $
+                     fromChunk <$> catMaybes (commonChunks ++ [remnant])
+
+    return (unserialised, serialised)
+    where
+
+    keyCount :: Int
+    keyCount = sum (map appendedKeysCount appends)
+
+    {-
+        We do not need to vary the minimum chunk size, since we are not testing
+        chunk building here.
+    -}
+    minChunkSize :: Int
+    minChunkSize = 0x1000
+
+-- * Properties to test
+
+-- ** Search
+
+prop_searchForMentionedKeyWorks :: NonNegative (Small Int)
+                                -> SearchableIndex
+                                -> Property
+prop_searchForMentionedKeyWorks (NonNegative (Small pageNo))
+                                (SearchableIndex index)
+    = pageNo < length lastKeys ==> all (<  key) prefix    .&&.
+                                   all (== key) selection .&&.
+                                   all (>  key) suffix
+    where
+
+    lastKeys :: Vector SerialisedKey
+    lastKeys = fromIndexOrdinary index
+
+    key :: SerialisedKey
+    key = lastKeys ! pageNo
+
+    prefix, selection, suffix :: Vector SerialisedKey
+    (prefix, selection, suffix) = searchPartitioning key index
+
+prop_searchForUnmentionedKeyInRangeWorks :: SerialisedKey
+                                         -> SearchableIndex
+                                         -> Property
+prop_searchForUnmentionedKeyInRangeWorks key (SearchableIndex index)
+    = key `notElem` lastKeys && key <= last lastKeys ==>
+      key < selectionHead                     .&&.
+      all (<  key)           prefix           .&&.
+      all (== selectionHead) (tail selection) .&&.
+      all (>  selectionHead) suffix
```

Comment:
I like the solution with adding the search partitioning to what is shown as the counterexample: simple, yet effective. It would indeed have saved me some work when investigating the bug fixed in #393. 😉 I implemented this solution in 7141645ae03955e250b0214949a0fa1177637a72.

I looked into using `conjoin` instead of `(.&&.)` but found the results not really convincing and therefore left things as they were in this regard.